### PR TITLE
fix(frontend): fix unpin_snapshot_before in HummockSnapshotManager

### LINKED ISSUE #5418: Investigate skip keys count too large in nexmark Q7
### Describe the bug

Investigate skip keys increase quickly. It seems to slow the throughput of our system.

image

image

### To Reproduce

run Nextmark Q7

### Expected behavior

The number of skip keys shall be stable

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hzxa21**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Prior to this PR, frontend will never unpin snapshot after `notify_epoch_assigned_for_queries` is called on empty batches, which causes compaction to fail to delete any stale key-value entries.

This PR fixes #5418.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
- https://github.com/risingwavelabs/risingwave/pull/5533

### CODE REVIEW

**Comment by hzxa21** [comment]
image

**Review by BowenXiao1999 — COMMENTED** [review]
LGTM.

BTW, if we do not want to maintain this complex unpin logic: we can simply send unpin request when receive epoch update.

**Comment by hzxa21** [comment]
> LGTM.
> 
> BTW, if we do not want to maintain this complex unpin logic: we can simply send unpin request when receive epoch update.

Agree. We can merge this PR to fix the issue first and simplify the current logic in future PRs.

**Review by BowenXiao1999 — APPROVED** [review]

**Review by Li0k — APPROVED** [review]
LGTM, thanks

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs [hunk 1]
```diff
@@ -278,11 +281,25 @@ impl HummockSnapshotManagerCore {
...
+                    epoch
+                );
+                *epoch
+            }
             None => last_committed_epoch,
         };
 
-        if min_epoch <= self.last_unpin_snapshot.load(Ordering::Acquire) {
+        let last_unpin_snapshot = self.last_unpin_snapshot.load(Ordering::Acquire);
+        tracing::info!(
```
Comment (Little-Wallace) [inline src/frontend/src/scheduler/hummock_snapshot_manager.rs:296]: This log info is too frequently....

**Review by Gun9niR — APPROVED** [review]
LGTM

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs (hunk 1 above)
Comment (hzxa21) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs:296]: > This log info is too frequently....

One log entry every UNPIN_INTERVAL_SECS (10sec) seems fine for me. Btw, we have a log entry for unpin rpc in L310.

**Comment by hzxa21** [comment]
Please hold this PR first. Found the following panic in my env:
```
2022-09-23T07:51:59.986194Z ERROR risingwave_storage::monitor::monitored_store: Failed in get: Hummock error: Expired Epoch: watermark 3059503322038272, epoch 3059503305588736.
  backtrace of inner error:
   0: <risingwave_storage::hummock::error::HummockError as core::convert::From<risingwave_storage::hummock::error::HummockErrorInner>>::from
             at ./src/storage/src/hummock/error.rs:63:10
   1: <T as core::convert::Into<U>>::into
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/convert/mod.rs:550:9
   2: risingwave_storage::hummock::error::HummockError::expired_epoch
             at ./src/storage/src/hummock/error.rs:117:9
   3: risingwave_storage::hummock::utils::validate_epoch
             at ./src/storage/src/hummock/utils.rs:58:20
   4: risingwave_storage::hummock::state_store::<impl risingwave_storage::hummock::HummockStorage>::read_filter
             at ./src/storage/src/hummock/state_store.rs:434:9
   5: risingwave_storage::hummock::state_store::<impl risingwave_storage::hummock::HummockStorage>::get::{{closure}}
             at ./src/storage/src/hummock/state_store.rs:300:13
   6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
   7: risingwave_storage::hummock::state_store::<impl risingwave_storage::store::StateStore for risingwave_storage::hummock::HummockStorage>::get::{{closure}}
             at ./src/storage/src/hummock/state_store.rs:451:69
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
   9: <async_stack_trace::StackTraced<F> as core::future::future::Future>::poll
  10: <futures_util::future::future::fuse::Fuse<Fut> as core::future::future::Future>::poll
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.23/src/future/future/fuse.rs:86:37
  11: <risingwave_storage::monitor::monitored_store::MonitoredStateStore<S> as risingwave_storage::store::StateStore>::get::{{closure}}
             at ./src/storage/src/monitor/monitored_store.rs:106:42
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
  13: risingwave_storage::keyspace::Keyspace<S>::get::{{closure}}
             at ./src/storage/src/keyspace.rs:103:76
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
  15: risingwave_storage::table::streaming_table::state_table::StateTable<S>::get_row::{{closure}}
             at ./src/storage/src/table/streaming_table/state_table.rs:409:22
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
  17: risingwave_stream::executor::managed_state::aggregation::value::ManagedValueState::new::{{closure}}
             at ./src/stream/src/executor/managed_state/aggregation/value.rs:57:63
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
  19: risingwave_stream::executor::managed_state::aggregation::ManagedStateImpl<S>::create_managed_state::{{closure}}
             at ./src/stream/src/executor/managed_state/aggregation/mod.rs:202:84
  20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
  21: risingwave_stream::executor::aggregation::generate_managed_agg_state::{{closure}}
             at ./src/stream/src/executor/aggregation/mod.rs:354:10
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
  23: risingwave_stream::executor::hash_agg::HashAggExecutor<K,S>::apply_chunk::{{closure}}::{{closure}}
             at ./src/stream/src/executor/hash_agg.rs:291:34
  24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
  25: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.23/src/stream/futures_unordered/mod.rs:514:17
```

Need to investigate more.

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs:296]: How about only log it when `min_epoch > last_unpin_snapshot`?

**Comment by hzxa21** [comment]
> Please hold this PR first. Found the following panic in my env:
> 
> ```
> 2022-09-23T07:51:59.986194Z ERROR risingwave_storage::monitor::monitored_store: Failed in get: Hummock error: Expired Epoch: watermark 3059503322038272, epoch 3059503305588736.
>   backtrace of inner error:
>    0: <risingwave_storage::hummock::error::HummockError as core::convert::From<risingwave_storage::hummock::error::HummockErrorInner>>::from
>              at ./src/storage/src/hummock/error.rs:63:10
>    1: <T as core::convert::Into<U>>::into
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/convert/mod.rs:550:9
>    2: risingwave_storage::hummock::error::HummockError::expired_epoch
>              at ./src/storage/src/hummock/error.rs:117:9
>    3: risingwave_storage::hummock::utils::validate_epoch
>              at ./src/storage/src/hummock/utils.rs:58:20
>    4: risingwave_storage::hummock::state_store::<impl risingwave_storage::hummock::HummockStorage>::read_filter
>              at ./src/storage/src/hummock/state_store.rs:434:9
>    5: risingwave_storage::hummock::state_store::<impl risingwave_storage::hummock::HummockStorage>::get::{{closure}}
>              at ./src/storage/src/hummock/state_store.rs:300:13
>    6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>    7: risingwave_storage::hummock::state_store::<impl risingwave_storage::store::StateStore for risingwave_storage::hummock::HummockStorage>::get::{{closure}}
>              at ./src/storage/src/hummock/state_store.rs:451:69
>    8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>    9: <async_stack_trace::StackTraced<F> as core::future::future::Future>::poll
>   10: <futures_util::future::future::fuse::Fuse<Fut> as core::future::future::Future>::poll
>              at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.23/src/future/future/fuse.rs:86:37
>   11: <risingwave_storage::monitor::monitored_store::MonitoredStateStore<S> as risingwave_storage::store::StateStore>::get::{{closure}}
>              at ./src/storage/src/monitor/monitored_store.rs:106:42
>   12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   13: risingwave_storage::keyspace::Keyspace<S>::get::{{closure}}
>              at ./src/storage/src/keyspace.rs:103:76
>   14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   15: risingwave_storage::table::streaming_table::state_table::StateTable<S>::get_row::{{closure}}
>              at ./src/storage/src/table/streaming_table/state_table.rs:409:22
>   16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   17: risingwave_stream::executor::managed_state::aggregation::value::ManagedValueState::new::{{closure}}
>              at ./src/stream/src/executor/managed_state/aggregation/value.rs:57:63
>   18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   19: risingwave_stream::executor::managed_state::aggregation::ManagedStateImpl<S>::create_managed_state::{{closure}}
>              at ./src/stream/src/executor/managed_state/aggregation/mod.rs:202:84
>   20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   21: risingwave_stream::executor::aggregation::generate_managed_agg_state::{{closure}}
>              at ./src/stream/src/executor/aggregation/mod.rs:354:10
>   22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   23: risingwave_stream::executor::hash_agg::HashAggExecutor<K,S>::apply_chunk::{{closure}}::{{closure}}
>              at ./src/stream/src/executor/hash_agg.rs:291:34
>   24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   25: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
>              at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.23/src/stream/futures_unordered/mod.rs:514:17
> ```
> 
> Need to investigate more.

This bug is caused by 

> Please hold this PR first. Found the following panic in my env:
> 
> ```
> 2022-09-23T07:51:59.986194Z ERROR risingwave_storage::monitor::monitored_store: Failed in get: Hummock error: Expired Epoch: watermark 3059503322038272, epoch 3059503305588736.
>   backtrace of inner error:
>    0: <risingwave_storage::hummock::error::HummockError as core::convert::From<risingwave_storage::hummock::error::HummockErrorInner>>::from
>              at ./src/storage/src/hummock/error.rs:63:10
>    1: <T as core::convert::Into<U>>::into
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/convert/mod.rs:550:9
>    2: risingwave_storage::hummock::error::HummockError::expired_epoch
>              at ./src/storage/src/hummock/error.rs:117:9
>    3: risingwave_storage::hummock::utils::validate_epoch
>              at ./src/storage/src/hummock/utils.rs:58:20
>    4: risingwave_storage::hummock::state_store::<impl risingwave_storage::hummock::HummockStorage>::read_filter
>              at ./src/storage/src/hummock/state_store.rs:434:9
>    5: risingwave_storage::hummock::state_store::<impl risingwave_storage::hummock::HummockStorage>::get::{{closure}}
>              at ./src/storage/src/hummock/state_store.rs:300:13
>    6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>    7: risingwave_storage::hummock::state_store::<impl risingwave_storage::store::StateStore for risingwave_storage::hummock::HummockStorage>::get::{{closure}}
>              at ./src/storage/src/hummock/state_store.rs:451:69
>    8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>    9: <async_stack_trace::StackTraced<F> as core::future::future::Future>::poll
>   10: <futures_util::future::future::fuse::Fuse<Fut> as core::future::future::Future>::poll
>              at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.23/src/future/future/fuse.rs:86:37
>   11: <risingwave_storage::monitor::monitored_store::MonitoredStateStore<S> as risingwave_storage::store::StateStore>::get::{{closure}}
>              at ./src/storage/src/monitor/monitored_store.rs:106:42
>   12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   13: risingwave_storage::keyspace::Keyspace<S>::get::{{closure}}
>              at ./src/storage/src/keyspace.rs:103:76
>   14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   15: risingwave_storage::table::streaming_table::state_table::StateTable<S>::get_row::{{closure}}
>              at ./src/storage/src/table/streaming_table/state_table.rs:409:22
>   16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   17: risingwave_stream::executor::managed_state::aggregation::value::ManagedValueState::new::{{closure}}
>              at ./src/stream/src/executor/managed_state/aggregation/value.rs:57:63
>   18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   19: risingwave_stream::executor::managed_state::aggregation::ManagedStateImpl<S>::create_managed_state::{{closure}}
>              at ./src/stream/src/executor/managed_state/aggregation/mod.rs:202:84
>   20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   21: risingwave_stream::executor::aggregation::generate_managed_agg_state::{{closure}}
>              at ./src/stream/src/executor/aggregation/mod.rs:354:10
>   22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   23: risingwave_stream::executor::hash_agg::HashAggExecutor<K,S>::apply_chunk::{{closure}}::{{closure}}
>              at ./src/stream/src/executor/hash_agg.rs:291:34
>   24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
>              at /rustc/9067d5277d10f0f32a49ec9c125a33828e26a32b/library/core/src/future/mod.rs:91:19
>   25: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
>              at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.23/src/stream/futures_unordered/mod.rs:514:17
> ```
> 
> Need to investigate more.

#5533 fixes this issue. We can merge this PR after #5533.

File: src/frontend/src/scheduler/hummock_snapshot_manager.rs (hunk 1 above)
Comment (hzxa21) [reply src/frontend/src/scheduler/hummock_snapshot_manager.rs]: > This log info is too frequently....

Changed to log it in trace level.

**Pull request merged**